sympy/matrices/common.py

```diff
@@ -2080,22 +2083,9 @@ def _eval_add(self, other):
         return self._new(self.rows, self.cols,
                          lambda i, j: self[i, j] + other[i, j])
 
-    def _eval_matrix_mul(self, other):
-        def entry(i, j):
-            try:
-                return sum(self[i,k]*other[k,j] for k in range(self.cols))
-            except TypeError:
-                # Block matrices don't work with `sum` or `Add` (ISSUE #11599)
-                # They don't work with `sum` because `sum` tries to add `0`
-                # initially, and for a matrix, that is a mix of a scalar and
-                # a matrix, which raises a TypeError. Fall back to a
-                # block-matrix-safe way to multiply if the `sum` fails.
-                ret = self[i, 0]*other[0, j]
-                for k in range(1, self.cols):
-                    ret += self[i, k]*other[k, j]
-                return ret
-
-        return self._new(self.rows, other.cols, entry)
+    def _eval_matrix_mul(self, other, mulsimp=None):
+        return self._new(self.rows, other.cols, lambda i, j: \
+                dotprodsimp(self[i,:], other[:,j], simplify=mulsimp))
```

Comment:
I think here you are creating a row and a column vector matrix with slicing. It means that you have an 1xn and a nx1 matrix. You are later assuming that you can iterate over these to get the elements. However numpy arrays iterate differently from sympy Matrices:
```julia
In [1]: M = Matrix([[1, 2], [3, 4]])                                                                                              

In [2]: from numpy import matrix                                                                                                  

In [3]: Mnp = matrix([[1, 2], [3, 4]])                                                                                            

In [4]: list(M[:,1])                                                                                                              
Out[4]: [2, 4]

In [5]: list(Mnp[:,1])                                                                                                            
Out[5]: [matrix([[2]]), matrix([[4]])]
```
This is coming from MatrixArithmetic which is part of the MatrixWrapper used after the numpy matrix is matrixified. I expect this is really a bug in the MatrixWrapper since it should produce something that behaves like a SymPy matrix.

Reply:
I've opened #17522